viewing_party/party.py

```diff
@@ -1,23 +1,217 @@
+    """
+    movie_dict = {}
+    if title and genre and rating:
+        movie_dict["title"] = title
+        movie_dict["genre"] = genre
+        movie_dict["rating"] = rating
+        return movie_dict
+
+    return None 
+
+
+def add_to_watched(user_data, movie):
+    """
+    input: user_data dictionary with nested list as values, movie dictionary
+    output: updated user_data dictionary 
+    """
+
+    user_data["watched"].append(movie) 
+
+    return user_data
+    
+
+def add_to_watchlist(user_data, movie):
+    """
+    input: user_data dictionary, and movie dictionary
+    output: updated user_data dictionary
+    """
+
+    user_data["watchlist"].append(movie) 
+
+    return user_data
+
+
+def watch_movie(user_data, title):
+    """
+    input: user_data dictionary, and title string
+    output: updated user_data dictionary 
+    """ 
+
+    for movie in user_data["watchlist"]:
+        if movie["title"] == title:
+            user_data["watchlist"].remove(movie)
+            user_data["watched"].append(movie)
+            return user_data
+    
+    return user_data
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    """
+    input: user_data dictionary with a "watched" list of movie dictionaries
+    output: average rating float
+    """
+
+    average_rating = 0.0
+    movie_ratings = [] 
+
+    for watched in user_data["watched"]:
+        rating = watched["rating"]
+        movie_ratings.append(rating)
+    
+    sum_of_movie_ratings = sum(movie_ratings)
+    length_of_movie_ratings = len(movie_ratings)
+    if length_of_movie_ratings == 0:
+        return average_rating
+    average_rating = sum_of_movie_ratings/length_of_movie_ratings
+
+    return average_rating
+
+
+def get_most_watched_genre(user_data):
+    """
+    input: user_data dictionary with a "watched" list of movie dictionaries
+    output: genre string
+    """
+
+    most_frequently_watched_genre = None
+    movie_genres = []
+
+    for watched in user_data["watched"]:
+        genre = watched["genre"]
+        movie_genres.append(genre)
+    if len(movie_genres) == 0:
+        return None
+
+    most_frequently_watched_genre = (mode(movie_genres))
+    return most_frequently_watched_genre
+
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    """
+    input: user_data dictionary with a "watched" list of movie dictionaries
+    output: list of dictionaries (which represents list of movies)
+    """
+
+    user_watched_list = []
+    friends_watched_list = []
+    movies_user_watched_that_friends_didnt_watch = []
+    
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched_list.append(movie)

```

Comment:
We can move this piece into a helper function as the next few functions need this same logic of "flattening" the friends' movies into a single list.